torch/utils/_cuda_trace.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from typing import Callable, Generic, List
+
+from typing_extensions import ParamSpec
+
+logger = logging.getLogger(__name__)
+P = ParamSpec("P")
+
+
+class CallbackRegistry(Generic[P]):
+    def __init__(self, name: str):
+        self.name = name
+        self.callback_list: List[Callable[P, None]] = []
+
+    def add_callback(self, cb: Callable[P, None]) -> None:
+        self.callback_list.append(cb)
+
+    def fire_callbacks(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        for cb in self.callback_list:
+            try:
+                return cb(*args, **kwargs)
```

Comment:
I think you added this `return` in error: this means the registry will only invoke the first callback in the list.

BTW, it could be a good idea to add a test that would have caught this bug :) (This is true in general: when finding a bug that wasn't caught by tests, it's common to add a new test for it, to catch regressions).

Reply:
You're correct. I've added a test checking if all callbacks are called.